specs/carbon-ql-k8s-specification.md

```diff
@@ -0,0 +1,357 @@
+  <key>: <value>
+model: # the model can be caculated locally or make a carbon-QL API remote call
+  path: <model name or Carbon-QL API URL path>
+  params :  # model params
+    <key>: <value>
+  telemetry : inline, or path, or queries (e.g Prometheus)
+  subComponents: # used to compose an SCI from sub components
+    <SubComponentKind>
+        - name: <another component same as above>
+          metadata : 
+          params:
+          path: 
+          telemetry:          
+  SCIAtrributedFrom: #used to attribute the SCI of the current component, based on the SCI of a hosting infra component (e.g shared resource)
+    <AttributedComponentKind>
+        - name: <another component same as above>
+          metadata : 
+          params:
+          path: 
+          telemetry:          
+```
+
+
+## Kubernetes components
+
+The Cabon-QL API includes the following components:
+
+Admin scenario:
+- Node (smallest infra component, aka foundational component)
+- Node Pool
+- Cluster
+
+Workload scenario:
+- Pod (smalled workload component, aka foundation component)
+- Deployment
+- Job
+
+## Kuberetnes components composition
+
+Carbon-QL components composition patterns for K8s:
+
+- **single foundational components (pod, node, disk,..)** : defines the base model to use, the params & telemetry, can set default vars
+
+```yaml 
+
+vars : <default vars>
+model:
+  path: azure.computte.aks.node.sci
+  params : <model params>
+  telemetry : <model telemetry>
+```
+ ---
+
+
+- **calculating SCI from subscomponents** : use sub components to calculate their SCI core, and defines the model to compute the resulting SCI (addition subcomponents SCI in most cases)
+
+```yaml 
+model:
+  path: gsf.common.AdditionSubComponentsSCIs:v1
+  params : ~
+  telemetry : ~
+  subComponents:
+    <SubComponentKind>
+        - name: <another component same as above>
+          metadata : 
+          params:
+          path: 
+          telemetry: 
+```
+
+the current components, when calling subcomponents, can override 
+          - metadata : 
+          - params:
+          - path: 
+          - telemetry: 
+
+
+
+----
+
+
+- **calculating SCI based on a attribution % from another SCI** : to calculate SCI for workloads and shared environments, an SCI for a pod, would be attributed  (or calculated) from the SCI of the hosting node (based on % of node resources consumed by the pod for the defined time duration)
+
+```yaml
+
+model:
+  path: azure.computte.aks.pod.sci
+  params:
+    <key>: <value>
+telemetry:
+  - <telemetry data>
+SubComponents: ~
+SCIAttributedFrom:
+  - SCIComponentKind: k8sNode (or point to SCIComponentName directly mynode1)
+    SCIComponentparams: 
+      nodeName : <node name>
+
+```
+
+
+
+## Scenarios
+
+## Example for Admin scenarios 
+
+An admin wants to calculate the infrastructure SCI for an AKS cluster in Azure. They can use the following components:
+
+- Node: my-node
+- Node Pool: my-node-pool
+- Cluster: my-cluster
+
+
+```yaml
```

Comment:
If we were mapping to the existing spec this is how I think we would write this, thoughts?

```yaml
name: AKS-PROD-1
tags:
  attributes: compute
  vendor: azure
  region: eastus
  sku: standard
  rg : rg-prod-1
  clusterSKU: vars.sku
  clusterName : name
  clusterResourceGroup : vars.rg
components:
  k8sNodePool:
    - name: mynode_pool-1
      tags:
          app: batch
      model: 
        path: azure.aks.node.sci:v1
      observations:
        - 2023-07-06T00:00:
          duration: 5
          cpu: 0.34
        - 2023-07-06T00:05:
          duration: 5
          cpu: 0.23
        - 2023-07-06T00:05:
          duration: 5
          cpu: 0.11        
    - name: mynode_pool-2
      model: 
        path: azure.aks.node.sci:v1    
      tags:
          app: web
      observations:
        - 2023-07-06T00:00:
          duration: 5
          cpu: 0.34
        - 2023-07-06T00:05:
          duration: 5
          cpu: 0.23
        - 2023-07-06T00:05:
          duration: 5
          cpu: 0.11
```